### ISSUE TITLE
Initial generics implementation

### ISSUE THREAD

**Issue by supersoftcafe** [supersoftcafe/yafl#3]
Angle brackets after a named thing provides type parameters, following a hybrid approach similar to Rust's traits and Haskell's classes, where a type is said to be allowable if it implements the requisite functions. This is a more useful approach than the Java/Kotlin/C# approach of declaring inheritance constraints on generic types.

```

# Just a boring interface
interface Stringable {
    fun toString(): String
}

# This trait is implicitly created from the interface declaration. We'll just put it in here for clarity.
trait Stringable<V> {
    fun toString(v:V): String
}

# This concrete implementation is implicitly created from the interface declaration. Again, added for clarity.
impl trait Stringable<Stringable> {
    fun toString(v:V): String => v.toString()
}

# Now for something that uses the above
fun print<V>(v:V): Unit where Stringable<V> => printString(toString(v))

# Let's extend Int32
impl trait Stringable<Int32> {
    fun toString(v:Int32): String => convertInt32ToString(v)
}

# Traits can implement (or extend depending on how you look at it) other traits
# Like Rust diamond paradox is dealt with by requiring each part of the hierarchy to be implemented separately.
trait MegaStringable<V> : Stringable<V> {
    fun doMega(v:V): Int32
}

# Traits aren't limited to a single target type. This one requires two type parameters.
trait ConvertBetween<X,Y> {
    fun convert(x:X):Y
}

# Finally, the style of generics we're used to from C#/Java. In the 'were' clause we use the colon
# syntax instead to show that we're limiting inheritance. Here we require both interfaces.
class BigContainer<X>(x:List<X>) where X : Stringable&Hashable {
}

# In conclusion.
# 1. Traits allow the definition of groups of functions on sets of types.
# 2. Generic constraints allow us to require that one more more types will have certain functions available and resolvable at compile type.
# 3. We can also dip into the Java/C# model of limiting class types to certain interfaces.


```

Don't forget:
* Static scoped methods. Need this to require 'new' as a constructor, for example.
* Instance scoped methods.
* Generic substitution on aliases.
